Feat/tabs accessibility : 접근성 속성 및 키보드 이벤트 추가

iyu88 opened the pull request:
## 🤠 개요

- #66 
- 접근성 속성을 추가하고 키보드 이벤트를 추가합니다. 


## 💫 설명

- MDN에서 정의한 탭의 role 속성을 Trigger에 추가했습니다. ( role="tab")

- 추가로 tabList와 tabPanel 역할도 List와 Panel에 추가했습니다. 

- Trigger가 Focus된 상태에서 Tab을 누르면 Panel로 이동해야 하기 때문에 선택된 Trigger의 tabIndex는 0으로, 
  그렇지 않으면 -1로 합니다.

- 다음의 aria-* 속성을 추가했습니다. 

  - aria-label : 어떤 탭 목록을 나타내는지 명시합니다.

    - 이를 위해서 List에서 label 속성을 props로 전달받습니다.  

  - aria-orientation : 탭으 방향을 명시합니다. 수평 탭만 지원하고 있기 때문에  horizontal 을 고정값으로 사용합니다. 

  - aria-selected : 선택된 Trigger 에서 true 값을 가집니다. 

  - aria-controls : Trigger가 제어하는 Panel의 id 값을 명시합니다. (value 로 Panel의 id를 설정했습니다.)

  - aria-labelledby : Panel을 제어하고 있는 Trigger의 id 값을 명시합니다. (value 로 Trigger의 id를 설정했습니다.)
 
- 키보드 방향키로 탭을 이동할 수 있게 했습니다. 

  -  constants/key.ts에서 이동방향이 같은 방향키들의 key 코드와 key 이름을 배열에 정의했습니다. 

  - Trigger가 Focus된 상태에서 다음 Trigger 혹은 이전 Trigger가 존재할 경우 해당 요소로 이동합니다. 

    - 이 때, selectedIndex 값도 변경하여 Panel 도 변경됩니다. 

- ScrollIntoView를 적용하여 이동한 Trigger가 화면에 표시되도록 합니다. 

  - ref 를 적용해야 해서 Button 컴포넌트를 사용하지 않고 button 태그를 사용합니다. 

  - 스크롤 이동을 확인하기 위한 스토리를 추가했습니다. 

---

+추가)

context에 대해서 계속 null 체크해주는게 불편해서 컨텍스트에 있는 값을 사용하는 훅을 선언했어요

```
const useTabs = () => {
  const context = useContext(TabsContext);
  if (context === null) {
    throw new Error('useTabs should be used within Tabs');
  }
  return context;
};
```
early return 할 때 어제는 fragment 를 반환하자는 입장이었는데, 저렇게 에러하는 레퍼런스를 봐버려서 일단 throw error 하도록 했습니다. 

```
const List = ({ children }: TabListProps) => {
  const { label, isFitted } = useTabs();
  // early return 필요 없음 
 // (생략)
```
실제 Compound 요소 안에서 사용하면 위와 같이 사용할 수 있어요.
이제 매번 early return 으로 처리하는 수고를 덜 수 있을 것 같아요.

참고했던 블로그 링크 남길게요. 읽어보시면 저보다 더 잘 활용하실 수 있을거라 믿어요~~ 

[다른 사람들이 안 알려주는 리액트에서 Context API 잘 쓰는 방법](https://velog.io/@velopert/react-context-tutorial#context-%EC%97%90%EC%84%9C-%EC%83%81%ED%83%9C-%EA%B4%80%EB%A6%AC%EA%B0%80-%ED%95%84%EC%9A%94%ED%95%9C-%EA%B2%BD%EC%9A%B0)

---

+추가)

Trigger 가 맨앞에 Focus된 상황에서는 맨끝으로, 맨끝에 Focus된 상황에서는 맨앞으로 이동해야 한다는 점을 알게 되어서 
함수에 동작을 추가했어요.

- 끝에서 끝으로 Focus

https://user-images.githubusercontent.com/31645195/227310577-15ecf90a-49a2-4387-ae1e-b9c196bce88a.mov

<br />

- Disabled 일 경우도 피해갑니다. (이 부분은 Disabled 를 커스텀하시면서 확인해보세요)

https://user-images.githubusercontent.com/31645195/227310569-7a2f2d24-4ecc-4ead-92e4-ca4454dd8f4f.mov

iyu88 committed:
feat: 접근성 관련 속성 추가

Trigger 와 Panel 을 aria-controls 와 aria-labelledby 속성으로 서로 이어주었습니다.
각 Compound 컴포넌트에 알맞은 role 속성을 부여합니다.
cds-tabs-[Component Name] 방식으로 id 를 추가합니다.

iyu88 committed:
feat: tabList 에 aria-label 추가

tabList 에 대해서 설명하는 DOM 이 없기 때문에 aria-label 속성을 사용합니다.
이를 위해 label 을 props 로 전달받습니다.

iyu88 committed:
feat: tabIndex 속성 추가

Trigger 에서는 현재 선택된 컴포넌트 이외에는 Tab 키의 영향을 받지 못하도록 -1 로 설정합니다.
Panel에서는 Trigger로부터의 Tab 키 이벤트를 받을 수 있도록 기본 값을 0으로 합니다.

iyu88 committed:
refactor: early return 값 Fragment 로 변경

iyu88 committed:
feat: Trigger 내부 Button 컴포넌트 제거

Button 컴포넌트를 사용하지 않고 기본 button 태그를 사용합니다.
ref 를 사용하여 Trigger 를 눌렀을 때 스크롤이 움직이도록 합니다.

iyu88 committed:
feat: 키보드 이벤트 등록

Trigger에 focus 된 상태에서 방향키를 움직이면 다음 Trigger 혹은 이전 Trigger 를 선택합니다.

iyu88 committed:
refactor: scrollOptions 분리

se030 commented on src/components/Tabs/index.tsx:
```diff
@@ -137,14 +141,63 @@ const Trigger = ({
-      text={children?.toString()}
+    <button
+      ref={triggerRef}
+      id={`cds-tabs-trigger-${value}`}
```
이렇게 고정값이 들어가면 id가 고유한 값임을 보장해주지 못하는 것 같아요. 사이드바가 탭 형식으로 되어있고 내부에서 탭을 또 사용하는 경우..?를 억지로 생각해봤어요.
그리고 접근성을 위해 제공하는 정보이니 tabs trigger보다 더 정확한 표현이 들어가는게 좋지 않을까요?
List에서 이미 label을 받고있으니까 Tabs props로 한 레벨 올려서 `${label}-trigger-${value}` / `${label}-panel-${value}`처럼 사용하면 어때요? 결국 TabList에 붙이고 싶은 라벨이 Tabs 전체에 대한 설명이기도 한 것 같아요.

iyu88 replied:
반영했습니다. 
접근성에 대해서 방향까지 제시해주셔서 너무 감사드려요!
다른 컴포넌트 만들때도 참고할게요~

se030 commented on src/components/Tabs/index.tsx:
```diff
@@ -137,14 +141,63 @@ const Trigger = ({
+      aria-selected={isActive}
+      aria-controls={`cds-tabs-panel-${value}`}
+      tabIndex={isActive ? 0 : -1}
       disabled={disabled}
```
aria-diabled도 넣어주세요!

se030 commented on src/components/Tabs/index.tsx:
```diff
@@ -137,14 +141,63 @@ const Trigger = ({
+
+    let futureTrigger;
+
+    if (NEXT_KEY.includes(e.key)) {
+      futureTrigger = currentTrigger.nextElementSibling as HTMLElement;
+    }
+
+    if (PREV_KEY.includes(e.key)) {
+      futureTrigger = currentTrigger.previousElementSibling as HTMLElement;
+    }
```
futureTrigger 가져올 때 disabled 여부도 확인해줘야 할 것 같아요!
아래쪽에서 focus 메서드는 동작을 안하고 그 뒤에 두 라인은 실행돼요.

iyu88 replied:
맞아요.. 이거 만들 때 해야지 했는데 안했네요.
futureTrigger를 받는 함수를 따로 분리했습니다. 
내부에서 while 문으로 disabled 면 다음 요소 찾도록 만들었어요~

se030 reviewed (commented):
고생하셨습니다! 현빈님꺼 반영되면 저도 상수 쓸래요 ㅎㅎ

iyu88 committed:
refactor: Id 및 접근성 속성 값 수정

기존의 Id가 고유성을 보장하지 못하는 것 같아 label 값을 사용하여 Id 를 부여합니다.
aria 속성 역시 Id 값을 활용합니다.
label props를 모든 Tabs 에서 공유하기 위해서 ContextProvider에서 props로 받습니다.
Trigger의 disabled 여부에 따라 aria-disabled 속성도 달라지도록 명시합니다.

iyu88 committed:
fix: findFutureTrigger 함수 추가

기존에 disabled 된 Trigger 도 키보드로 이동할 수 있었던 점을 수정합니다.
동시에 FutureTrigger 를 찾는 함수를분리합니다.

iyu88 committed:
fix: 스토리 label 추가

iyu88 committed:
refactor: useTabs 커스텀 훅 추가

Compound 요소에서 계속 context 의 null 체크하는 점이 불편하기 때문에 useTabs 커스텀 훅을 선언합니다.
더 이상 early return 으로 fragment 를 반환하지 않습니다.

iyu88 commented:
PR 메세지 하단에 새로운 내용을 추가했어요.
리뷰 전에 한 번씩 읽어주세요~ 🙇‍♂️

se030 commented:
> PR 메세지 하단에 새로운 내용을 추가했어요. 리뷰 전에 한 번씩 읽어주세요~ 🙇‍♂️

에러 안던지는걸로 정해진줄 알았는데 핸들링이 필요할 것 같다는 의견은 스킵인가요?

1. 네이밍은 useTabs보다 더 역할이 드러났으면 좋겠어요 ~ useTabsContext 어떤가요?

2. 사용방식을 이렇게 통일하고 싶다면 공통 컨텍스트 훅을 만들어두는게 좋을 것 같아요. useAbsoluteContext() 뭐 이런식으로 ..ㅎㅎ 제네릭 사용해서 리턴타입 명시하고요!

<br />

> 에러 안던지는걸로 정해진줄 알았는데 핸들링이 필요할 것 같다는 의견은 스킵인가요?

이거 관련해서는 에러 바운더리로 처리가 가능한데, 베타 문서에서 레거시 API로 분리된 클래스 컴포넌트 생명 주기를 사용할 수밖에 없다는게 조금 걸리는 부분이에요. 근데 현재 버전에서 따로 권장하는 에러 핸들링 방식이 없어서 공통 에러캐치가 필요하다면 이게 최선이라고 생각해요.

> There is no direct equivalent for componentDidCatch in function components yet. If you’d like to avoid creating class components, write a single ErrorBoundary component like above and use it throughout your app. Alternatively, you can use the [react-error-boundary](https://github.com/bvaughn/react-error-boundary) package which does that for you.

문서에서도 하나 만들어서 공통으로 사용하라고 얘기하고요. (함수 컴포넌트에서 유사한 기능이 생겼을 때 해당 모듈만 수정할 수 있어서인 것 같아요.) 저 [라이브러리](https://github.com/bvaughn/react-error-boundary/blob/master/src/ErrorBoundary.ts)도 열어보면 결국 클래스 컴포넌트 쓰는거라서 직접 만드는게 좋을 것 같아요. 컨텍스트 없을 때 에러 던지기로 한다면 저 만들어보고싶어요 ~

https://ko.reactjs.org/docs/error-boundaries.html
https://react.dev/reference/react/Component#componentdidcatch

iyu88 commented:
@se030 
1. 네이밍은 제안해주신대로 수정해볼게요!

2. 하긴 계속 반복될테니까 공통 컨텍스트 훅을 만들어보는 것도 좋겠네요. 별도로 이슈 하나 파서 작업하면 좋을 것 같아요 ㅎㅎ
에러는 일단 블로그 글 참고해서 작성해보았는데,
생각해보니까 저번에 회의할 때 잠깐 이야기 나왔던 것처럼
개발하는 과정에서 Component의 Context를 잘못 사용했을 경우에 _'에러 메세지가 바로 나와서 사용처가 잘못되었다는 걸 알려주는 목적'_
이라면 저대로 사용해도 괜찮을 것 같아요.

일단은 이미 결론이 난 문제이기 때문에 그냥 fragment 를 반환하는 방법을 사용하고, 
추후에 공통적인 Context Hook을 만들게 된다면 말씀하신대로 Error - Error Boundary 처리도 고려해보는게 어떨까요? 

결론은
1) 훅에서 fragment 반환 
2) 추후 공통 Context Hook 작업
3) (선택) Error Boundary 처리 

이렇게요!

se030 commented on src/components/Tabs/index.tsx:
```diff
@@ -137,15 +171,65 @@ const Trigger = ({
+
+    if (!(currentTrigger instanceof HTMLElement)) return;
+
+    let futureTrigger;
+
+    if (NEXT_KEY.includes(e.key)) {
+      futureTrigger = findFutureTrigger('NEXT', currentTrigger);
+    }
+
+    if (PREV_KEY.includes(e.key)) {
+      futureTrigger = findFutureTrigger('PREV', currentTrigger);
+    }
```
현빈님 이 부분 futureTrigger = findFutureTrigger(e.key) 처럼 변경가능할 것 같아요. 로직이 findFutureTrigger 함수의 역할이라는 생각이 들어요.
그리고 key 사용하는거면 상수에 37 38 같은 keycode는 없어도 괜찮을 것 같은데 혹시 예외상황이 있나요?

iyu88 replied:
말씀해주신대로 key 방향을 판단하는 로직까지 `findFutureTrigger()` 내부로 위치시켰습니다.
key 같은 경우에는 같은 방향키에 대해서 key, keycode 구분 없이 편하게 사용하면 좋을 것 같아서 포함시켰습니다!

se030 commented:
> @se030
> 
> 1. 네이밍은 제안해주신대로 수정해볼게요!
> 2. 하긴 계속 반복될테니까 공통 컨텍스트 훅을 만들어보는 것도 좋겠네요. 별도로 이슈 하나 파서 작업하면 좋을 것 같아요 ㅎㅎ
>    에러는 일단 블로그 글 참고해서 작성해보았는데,
>    생각해보니까 저번에 회의할 때 잠깐 이야기 나왔던 것처럼
>    개발하는 과정에서 Component의 Context를 잘못 사용했을 경우에 _'에러 메세지가 바로 나와서 사용처가 잘못되었다는 걸 알려주는 목적'_
>    이라면 저대로 사용해도 괜찮을 것 같아요.
> 
> 일단은 이미 결론이 난 문제이기 때문에 그냥 fragment 를 반환하는 방법을 사용하고, 추후에 공통적인 Context Hook을 만들게 된다면 말씀하신대로 Error - Error Boundary 처리도 고려해보는게 어떨까요?
> 
> 결론은
> 
> 1. 훅에서 fragment 반환
> 2. 추후 공통 Context Hook 작업
> 3. (선택) Error Boundary 처리
> 
> 이렇게요!


요거 첫번째가 아마 훅으로 분리하면 안될 것 같은데 !!

저는 좀 더 생각해보니 혹시 생길지 모를 misuse에 대해서는 컴포넌트를 렌더링하지 않는 것보다는 에러 던지는게 좋아보여요.
사실 개발자에게 보여주기 위한 에러면 캐치 안해도 ..ㅎㅎ 되지 않을까 싶은데 Error Boundary를 사용한다면 목적은 딱 콘솔 에러로그를 UI로 보여주는 용도?
우선 만들어두신 훅 그대로 두고 다른 컴포넌트들은 순차적으로 수정하면 어떨까요? 다른 분들도 의견 주세요 ㅎㅎ

iyu88 commented:
> 사실 개발자에게 보여주기 위한 에러면 캐치 안해도 ..ㅎㅎ 되지 않을까 싶은데

네네 좋아요!
Compound 패턴을 위한 Context를 사용하게 된다는 점을 미루어 보았을 때,
결국 사용자 (개발자)가 컨텍스트 커스텀 훅을 
전역 상태가 공유되고 있는 위치에서만 사용해야하니까 
그렇지 않은 경우에 잘못되었다는 에러를 던지는게 좋을 것 같네요~

iyu88 committed:
refactor: useTabsContext 커스텀 훅 이름 변경

prayinforrain reviewed (commented):
고생하셨어요~! 사실 이전부터 계속 보고있었는데  
링크에 첨부하신 글 읽다가 중간에 뭐 일생겨서 딴거 하다가 또 다시 글 읽다가 하느라 늦었어요.. 😅  
  
우선 context 관련된 훅 공통으로 사용하는거 저는 너무 찬성이에요. 이번 주에 다들 컨텍스트 많이 사용하셨을 것 같은데 일관성 지키기도 좋고 유지보수도 편해질 것 같아요. 빨리 적용해야해요 ㅋ ㅋ

그런데 직접 Tabs를 사용해보니 조금 걸리는 점이 두 가지 있어요. 한번 읽어 보시고 다른 의견 있으시면 제시해 주세요!
- 보통 키보드로 탐색할 때에는 Tab으로 focus를 움직이고, 열어야 할 때 Enter나 Space를 눌러서 active하는 프로세스라고 생각해요. 그런데 지금은 focus를 움직이는 동시에 active 작동이 함께 이루어지니까, 만약 1번 탭을 보다가 3번 탭으로 이동해야 한다면 2번 탭을 불필요하게 거쳐야 하는 문제가 생길 것 같아요.
- 선택되지 않은 탭의 tabIndex를 -1로 하면 안될 것 같아요. 개인적으로 풀 키보드 브라우징을 하지는 않지만 가끔 탭을 사용할 때가 있는데, 탭을 선택한 후에도 탭을 한 번 더 누르면 그 옆 탭이 focus되지, Panel로 focus가 이동하진 않았던 것 같아요. 실제로 접근성 도구가 어떻게 작동하는지는 모르겠지만 제 좁은 경험 상으로는 조금 이질적인 동작인 것 같습니다.

예시를 찾고 싶어서 CSR이면서 탭 컴포넌트가 있는 페이지를 찾고 있는데 떠오르는게 잘 없네요. 아무튼 제 의견은 이렇습니다 🤔

iyu88 committed:
refactor: findFutureTrigger 내부에서 key 방향 판단

iyu88 committed:
chore: Trigger 텍스트 변경

iyu88 commented:
말씀해주신 부분들에 대해서 [MDN - tablist_role](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/tablist_role)을 토대로 답변드리면,

> 1. 보통 키보드로 탐색할 때에는 Tab으로 focus를 움직이고, 열어야 할 때 Enter나 Space를 눌러서 active하는 프로세스라고 생각해요. 그런데 지금은 focus를 움직이는 동시에 active 작동이 함께 이루어지니까, 만약 1번 탭을 보다가 3번 탭으로 이동해야 한다면 2번 탭을 불필요하게 거쳐야 하는 문제가 생길 것 같아요.

우재님이 말씀해주신대로 Enter 와 Space 의 역할에 대해서 따로 명시하고 있어요.
> Activates the tab if it was not activated automatically on focus.

focus 했을 때 자동으로 활성화되지 않았을 경우 Enter / Space를 눌러서 활성화하라는 의미로 해석했어요. 

여기서 잠깐 화살표 키 (Left Arrow) 를 보면,
> Optionally, activates the newly focused tab

새로 focus 된 탭을 활성화하라는 옵션이 있었어요.

따라서 위 옵션을 적용하면 Enter나 Space 에 대한 처리를 하지 않아도 될 것이라 판단했어요.

추가적으로 말씀해주신 내용처럼

**그냥 지나가는 탭인데 불필요하게 focus & active 되어 Panel이 변경되는 불편함**을 해결하기 위해 Enter를 도입하는 것은

반대로 **모든 탭을 다 확인하고 싶어서 Left Arrow + Enter 를 반복해야 하는 불편함** 을 만들어 낼 수 있다고 생각해요.

정답이 없는 문제 같아요. 이 부분에 대한 다른 분들의 의견도 부탁드려요. (혹시 제가 잘못 이해한 부분이 있으면 알려주세요..!!!)

---

> 2. 선택되지 않은 탭의 tabIndex를 -1로 하면 안될 것 같아요. 개인적으로 풀 키보드 브라우징을 하지는 않지만 가끔 탭을 사용할 때가 있는데, 탭을 선택한 후에도 탭을 한 번 더 누르면 그 옆 탭이 focus되지, Panel로 focus가 이동하진 않았던 것 같아요. 실제로 접근성 도구가 어떻게 작동하는지는 모르겠지만 제 좁은 경험 상으로는 조금 이질적인 동작인 것 같습니다.

이 부분은 

_Trigger 가 Focus된 상태에서 Tab을 누르면 Panel 이 Focus 되는 것보다 다음 Trigger가 Focus 되는 것이 자연스럽다_ 

라고 이해했어요.

링크에 있는 Tab 설명을 보면 

> When the tab list contains the focus, moves focus to the next element in the page tab sequence outside the tablist, which is the tabpanel unless the first element containing meaningful content inside the tabpanel is focusable.

tablist에서 Tab을 누르면 tablist 바깥인 panel로 이동하고, panel 안에 focus 가능한 요소가 있으면 해당 요소로 이동하라는 의미로 해석하여 panel 로 이동하도록 ( = 다음 Trigger로 이동하지 않도록 ) 다른 Trigger 들의 tabIndex 를 -1로 설정했어요.

여기에 대해서 제 해석이 틀렸다면 알려주세요!!

---

+ 접근성 문서를 한 번 더 읽어보게 되었는데, 선민님 Carousel 하신 것처럼 Tab도 무한 스크롤(?) 이 되어야 하네요.

작업이 늘었어요 ^_^

> moves focus to the previous tab. If focus is on the first tab, moves focus to the last tab. - Left Arrow 중 -

iyu88 committed:
feat: Trigger Focus 가 양끝에 있을 경우 화살표 이동 처리

MDN 문서에서 Focus 된 Tab 이 맨처음이나 맨끝에 있을 경우 각각 맨끝과 맨처음으로 이동해야 한다는 점을 함수에 추가합니다.
이 때, 새롭게 Focus한 요소가 disabled 상태인지 확인합니다.

prayinforrain reviewed (approved):
지금 보니 첫 번째 문제는 사실 탭 전환이 해당 탭의 내용을 추가로 렌더링하게 되지 않나 하는 생각에(예를들면 data fetching이나 useEffect같은 로직을 탭을 열었을 때 실행하는 줄 알았어요) 이상하다고 느낀건데 이 코드 상에서 생각하면 있을 수 없는 문제네요. 너무 복잡하게 생각했나봐요..  

그럼에도 제 개인적인 선호는 Enter로 confirm을 거치는 게 좋다고 생각하는데, [mdn 예제](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/tab_role#example)는 그렇게 되어 있고 지금 제가 쓰고 있는 GitHub 에디터는 현빈님이 구현해 주신 대로 동작하고 있네요. 말 그대로 이래도 좋고 저래도 좋은 사항인 것 같아요.

>  For controlling tab key functionality for elements with the role tab, it is recommended to set all non-active elements to tabindex="-1", and to set the active element to tabindex="0".

두 번째는 [Best practice](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/tab_role#best_practices)라고 명시가 되어있는 사항이였네요. 역시 찾아보면 다 나오는건데 너무 제 기준으로 넘겨짚었어요 😅

접근성 정말 끝이 안나네요 고생하셨습니다 👍👍

iyu88 committed:
refactor: 이벤트 핸들러 네이밍 컨벤션 준수

handle 로 시작하던 이벤트 핸들러 네이밍을 onVerb 유형으로 수정합니다.

se030 reviewed (approved):
고생하셨습니다 👏🏻👏🏻👏🏻👏🏻

우재님이 코멘트 해주신 부분에 대해서는, 스펙을 어떻게 결정할건지의 문제라고 생각했어요.
[이 링크](https://nuli.navercorp.com/community/article/1133145)를 읽고 바닐라로 따라해 본 적이 있는데, 여기서는 auto-select / manual-select 두개의 모드로 구현하셨더라고요. 같은 컴포넌트라도 접근성 동작은 여러가지를 지원할 수 있고, 접근성에만 한정되는 이야기도 아닌 것 같아요. 형태나 기본 동작들도 너무 다양하니까요.

캐러셀도 그렇고 기획이나 스펙이 명확하지 않은 상태이다보니 각자가 생각하는 구현체가 달라서 작은 혼란이 생긴다고 느껴져요.
다음부터는 구현 전에 설계 회의같은거 한번씩 하면 어떨까요? 다같이 하면 시간이 오래 걸리니까 두세명씩 서로 설계 봐주고 의견 모아서 방향 정하면 더 효율적으로 작업할 수 있을 것 같아요.

iyu88 committed:
refactor: List 에서 isFitted 값을 사용하지 않도록 변경

List에서 사용하고 있던 isFitted 도 결국 Trigger 와 관련된 값이기 때문에 Trigger 에서 직접 가져와서 CSS 속성을 설정하는 데에 사용합니다.
